### PR TITLE
Fix quantization in the effect engine (metronome effect)

### PR DESCRIPTION
**Pull request by daschuer**
This was originally a fix for #13611 but it turns out that the beat information provided into the effect engine itself was broken. I have fixed this at the root by refactoring GroupFeatureState and the "beat_next" and "beat_prev" COs.
Than I was able to finally fix the metronome, which works now also reliable during scratching.

Not sure if there is a real use case for it, but the metronome was perfectly to test that, and also helps verifying a beat grid by ear. It would be a real fun to play any sample and not only the metronome click. That could become our first "transport" effect. I will file a feature request. 

Open is to check/fix other effects. I want to have that reviewed and merged first before building on that. 
See commits for details.

### CODE REVIEW

File: src/effects/backends/builtin/metronomeeffect.cpp [hunk 1]
```diff
@@ -77,9 +77,8 @@ void MetronomeEffect::processChannel(
     }
 
     SINT maxFrames;
-    if (m_pSyncParameter->toBool() && groupFeatures.has_beat_length_sec) {
-        maxFrames = static_cast<SINT>(
-                engineParameters.sampleRate() * groupFeatures.beat_length_sec);
+    if (m_pSyncParameter->toBool() && groupFeatures.has_beat_length_frames) {
+        maxFrames = static_cast<SINT>(groupFeatures.beat_length_frames);
```
[the remark this answers is not in the document]
Comment (daschuer) [reply src/effects/backends/builtin/metronomeeffect.cpp]: I have replaced it by double precision in a later commit.

**Comment by daschuer** [comment]
Done.

File: src/effects/backends/builtin/metronomeeffect.cpp
Comment (Swiftb0y) [inline src/effects/backends/builtin/metronomeeffect.cpp:1]: the changes here are really hard to follow because they mix offset and timing calculations with buffer indexing (mixing levels of abstraction) and thus I'm neither confident about the reliability nor safety of the code...

Do you think its feasible to reorganize it a little so its easier to understand?

Comment (daschuer) [reply src/effects/backends/builtin/metronomeeffect.cpp:1]: I have tried to fix that by calculate everything in frames. Maybe I was not too successful or comments are missing. 
Can you give some hints what can be improved?

Comment (Swiftb0y) [reply src/effects/backends/builtin/metronomeeffect.cpp:1]: I will rework it a little and open a PR to your branch ASAP.

Comment (Swiftb0y) [reply src/effects/backends/builtin/metronomeeffect.cpp:1]: Please take a look https://github.com/daschuer/mixxx/pull/97

**Comment by daschuer** [comment]
I have added the first commit of your branch from https://github.com/daschuer/mixxx/pull/97
Can we merge this first, before doing the std::span refactoring? 
I am not convinced that this reduces complexity, because it brings back the sample count where I am very happy to with the frame based counters.

**Comment by Swiftb0y** [comment]
> Can we merge this first, before doing the std::span refactoring?

Sure, but unfortunately I encountered a couple bugs during testing:

1. Sync on the Metronome is off: expected it should continue clicking based on the fixed rate of the knob. Currently Observed: it only does not click at all, only when enabling the effect, and with its transient chopped of.
2. Sync on the Metronome is on but track has no beatgrid. Expected: not quite sure actually whether it should stay silent or just play as if sync was turned off. Observed: it just clicks once when enabling, though this time its the full sample.

Both of these issues are regressions.

**Comment by daschuer** [comment]
The described issues are fixed now. It was caused by a faulty fix to avoid SINT overflow.

**Comment by Swiftb0y** [comment]
I don't understand the fix... There also remains the issue where the first click is with sync disabled is muffled.

**Comment by daschuer** [comment]
> There also remains the issue where the first click is with sync disabled is muffled.

This is a "feature", the fade in of the effect. But we can safely disable this here. 

I have removed a redundant check and added comments to the last summit.

File: src/effects/backends/builtin/metronomeeffect.cpp [hunk 2]
```diff
@@ -57,71 +58,87 @@ void MetronomeEffect::processChannel(
...
+                        beatLength * -1 *
+                        (1 - *groupFeatures.beat_fraction_buffer_end);
+            }
+
+            if (bufferEnd > beatToBufferEnd) {
+                // We have a new beat before the current buffer ends
+                nextClickStart = bufferEnd - beatToBufferEnd;
+            }
+        } else {
+            nextClickStart = *groupFeatures.beat_length_frames;
```
Comment (Swiftb0y) [inline src/effects/backends/builtin/metronomeeffect.cpp:125]: Can this actually happen?

Comment (daschuer) [reply src/effects/backends/builtin/metronomeeffect.cpp:125]: No, I have added it only for completeness. We may consider to use both in a optional struct to get around it.

Comment (Swiftb0y) [reply src/effects/backends/builtin/metronomeeffect.cpp:125]: do you mean `beat_length_frames` and `beat_fraction_buffer_end` or `beat_fraction_buffer_end` and `scratch_rate`? because according to `BpmControl::collectFeatures` if `rate_ratio` is 0, we can get a `beat_fraction_buffer_end` but not a `beat_length_frames`

Comment (daschuer) [reply src/effects/backends/builtin/metronomeeffect.cpp:125]: We always have it all in reality, because we do not allow a rate of 0 but it is possible via writing the CO. 
We should probably treat this case in the same way we treat scratch = 0

Comment (Swiftb0y) [reply src/effects/backends/builtin/metronomeeffect.cpp:125]: yes, I agree. I implemented that in a follow up. will push shortly.

Comment (daschuer) [reply src/effects/backends/builtin/metronomeeffect.cpp:125]: I am already on it.

**Comment by daschuer** [comment]
Done, please have a look.

File: src/effects/backends/builtin/metronomeeffect.cpp (hunk 2 above)
Comment (Swiftb0y) [reply src/effects/backends/builtin/metronomeeffect.cpp]: sure that works, but it doesn't eliminate the original case I asked about (which is when `beat_length_frames.has_value()` but `!beat_fraction_buffer_end.has_value()`).

Comment (daschuer) [reply src/effects/backends/builtin/metronomeeffect.cpp]: This can not happen. But looking at the API it, can. `beat_fraction_buffer_end.has_value() == false` means the track has no value. In this case `beat_length_frames.has_value()` is always false.

Comment (Swiftb0y) [reply src/effects/backends/builtin/metronomeeffect.cpp]: right, so that should be reflected in `GroupFeatureState`, do you agree?

**Comment by Swiftb0y** [comment]
fyi, I got https://github.com/daschuer/mixxx/pull/97 into a state I'm happy with. Please take a look.

File: src/effects/backends/builtin/metronomeeffect.cpp (hunk 2 above)
Comment (daschuer) [reply src/effects/backends/builtin/metronomeeffect.cpp]: I am not sure, I because we may consider a sound source that has bpm info but no phase info.

**Comment by daschuer** [comment]
I have adopted the full span based approach. But I have not adopted the rest, because of the std::size_t / std::ptrdiff_t / SINT clash with gsl/span. I also must admit that I can not fully understand how your code works.

**Comment by Swiftb0y** [comment]
I think the primary advantage of my span approach was not the fact that it uses span (it just made it easier for myself to do the cleanup I wanted to do). The real improvement was separating the messy sync code from the rest. But I'm okay with the current state as a temporary stopgap.

**Review by Swiftb0y — APPROVED** [review]
This LGTM for now. I will retarget the follow up at `main`.
